policy: Add support for tiers and pass verdicts

jrajahalme opened the pull request:
Add support for arbitrary number of policy tiers and pass verdicts.

This is work-in-progress. The following items are still to be done:
- [ ] always insert a default rule so that admin tier rules after a PASS verdict will always be bypassed
- [ ] Envoy policies do not yet support policy tiers or pass verdicts

```release-note
Policy engine adds support for tiers and pass verdicts for ClusterNetworkPolicies
```

jrajahalme committed:
policy: Change PerSelectorPolicy.IsDeny bool to Verdict enum

Change PerSelectorPolicy to use Verdict enum instead of Change IsDeny
boolean. Zero value for the Verdict enum is 'Invalid' so an explicit
'Allow' verdict must be placed in PerSelectorPolicies.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
policy: Add support for Tiers and Pass verdicts

Add support for tiers and pass verdicts in policy computation. Neither is
implemented in the policy APIs yet.

Mapstate entries are inserted by tier, starting with the highest tier
(0). This ensures that higher tier pass verdicts are present in mapstate
when inserting lower-tier entries.

If an entry added in a lower tier is covered by a higher-tier entry with
PASS verdict, the precedence of the added entry is elevated to the
precedence of the PASS verdict. This precedence elevation is needed to
have the "passed-to" verdict for a specific identity to take precedence
over possible higher tier wildcard identity verdicts. When ingesting
policy rules a gap in priority space is left after each pass verdict to
allow for the priorities needed for the rules of all the lower tiers.

An allow or deny policy key can be the same as the covering pass policy
key on a higher tier. This requires managing pass related information as
metadata rather than a 3rd verdict in the MapStateEntry. This lines up
well with the fact that keys with pass entries are not synced to the bpf
datapath. The pass metadata consists of two new fields in mapStateEntry:

  - passPrecedence is the precedence of the pass entry itself, or 0
    if none.
  - nextTierPrecedence stores the base precedence of the next tier, used
    in elevation of the priority of the passed-to entries to start with,
    or immediately follew the passPrecedence

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
policy: Factor deleteKeyWithChanges into deleteExistingWithChanges

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
policy: Make PortRules a slice of L4PolicyMap

Prepare for policy tier support by making PortRules a slice of
L4PolicyMap, one for each tier. There is always at least one tier for
legacy support.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
policy: ignore precedence in deleteKeyWithChanges

After deleteExistingWithChanges() refactor deleteKeyWithChanges() is only
used in an incremental delete of an identity. When deleting an identity,
all keys with that identity are removed, so we do not need to check that
the precedence matches. This prepares for pass verdict support where
mapstate entry precedence can be adjusted due to a pass verdict, so it
would not be possible to match the precedence any more without
recomputing the adustments.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
policy: Change PolicyEntry.Deny bool to Verdict enum

Change policy/types.PolicyEntry to use the Verdict enum instead of a
'deny' boolean. Zero value for the Verdict enum is 'Invalid' so an
explicit 'Allow' verdict must be placed in PolicyEntries.

A panic in non-production code ('repository.addListLocked()', called via
'repository.MustAddPolicyEntries()') fires if a rule has an invalid
verdict.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme commented:
Added commit to change PolicyEntry `Deny` to Verdict, but no-one puts a pass verdict there yet.

jrajahalme commented:
rebased and cleaned up, un-drafting.

squeed commented on pkg/policy/mapstate.go:
```diff
@@ -919,9 +943,11 @@ type MapChanges struct {
-	Key   Key
-	Value mapStateEntry
+	Add            bool // false deletes
+	Tier           types.Tier
```
In fact, shouldn't the Value have everything we need?

jrajahalme commented:
Restarted flow interruption test failure in `ci-l7`, could happen due to policy issue, but seems unlikely this PR would be causing it.

TheBeeZee commented on pkg/policy/types/entry.go:
```diff
@@ -144,15 +169,15 @@ func NewMapStateEntry(
-func (e MapStateEntry) Level() uint32 {
-	return MaxLevel - uint32(e.Precedence>>PrecedenceLevelShift)
-}
+//func (e MapStateEntry) Level() Precedence {
```
Can this be removed?

jrajahalme replied:
Changed the `String()` method to use this, so no :-)

TheBeeZee commented:
I ran this PR through the grinder. I applied my KCNP patch on top, and ran the KCNP conformance tests. The "integration" test fails. The test is as follows:

1. Admin tier policy that applies to traffic from ns=a to ns=b with PASS verdict.
2. Baseline tier policy that applies to traffic from ns=a to ns=b with DENY verdict.

Right now, this unexpectedly allows the traffic from ns=a to ns=b. Looking at the realized network policy that Cilium has, it seems to only keep the Admin tier policy, completely ignoring the Baseline tier policy. I think this might be because the policies match exactly (except for the verdict) and the Baseline policy due to lower priority gets pruned too early.

jrajahalme committed:
policy: Defer adding the wildcard entry

Only create the wildcard entry before looping over the cached selectors
(so that we know it exists), but add it when the cached selector is
iterated.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme commented on pkg/policy/types/policyentry.go:
```diff
@@ -8,21 +8,32 @@ import (
 	// take precedence over higher values. Rules having the default priority level 0 are
 	// considered first, then the rest of the rules, from the earliest to later priority levels.
-	// This is currently limited to 24 bits, i.e., max allowed priority is (1<<24-1).
 	Priority float64
```
It avoids a numbering problem when we want to add a new entry with priority between two existing entries with consecutive integer priorities (e.g., 0, and 1).

jrajahalme commented on pkg/policy/types/policyentry.go:
```diff
@@ -8,21 +8,32 @@ import (
+	Admin Tier = iota
+	Normal
+	Baseline
+	numTiers // one past the lowest tier
```
Changed to insert the default-allow wildcard rule on a tier of it's own, but only if there are more than 1 tiers or there are rules with non-zero priorities. This is for backwards compatibility and safety in policies that do not use these features.

jrajahalme commented on pkg/policy/l4.go:
```diff
@@ -1467,6 +1469,11 @@ func (pf policyFeatures) contains(feature policyFeatures) bool {
 type L4DirectionPolicy struct {
 	PortRules L4PolicyMaps
 
+	// TierPriority stores the starting priority for each tier.
```
Fixed, thanks!

jrajahalme committed:
policy: More wildcard selector testing

Add TestL3Wildcarding test case that covers L3 wildcarding with an
explicit wildcard selector, nil L3, and non-nil but empty L3 with
specific ports, wildcard ports, and no ports.

Update CRD docs accordingly.

Update the note on PolicyEntry.L3 that has different semantics:
- Nil selects nothing
- non-nil but empty is implicitly treated as a wildcard selector if any
  L4 PortRules are present.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
policy: Add policy/types.Verdict

Define policy/types.Verdict enum, use this instead of boolean 'deny' in
the internal mapstate API.

MapStateEntry.String() method is changed to translate precedence field to
API-level Priority and ListenerPriority types for more meaningful
debugging.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

sayboras reviewed (approved):
Looks good for proxy related changes

jrajahalme commented on pkg/policy/rules.go:
```diff
@@ -5,66 +5,158 @@ package policy
 
-	// sort rules (in place) by priority
-	rules.sort()
+func ensureSlice[E any, S ~[]E, I ~uint | ~uint8](s *S, index I) {
```
For most calls the allocation and copy are not needed, while the one above does it every time?

jrajahalme commented on pkg/policy/repository.go:
```diff
@@ -446,27 +451,49 @@ func (p *Repository) computePolicyEnforcementAndRules(securityIdentity *identity
 
 // wildcardRule generates a wildcard rule that only selects the given identity.
-func wildcardRule(lbls labels.LabelArray, ingress bool) *rule {
+func wildcardRule(lbls labels.LabelArray, ingress bool, tier types.Tier, priority float64) *rule {
```
Adding a test case `"PASS 1.1.1.1 over deny with default allow"` to make sure default allow rules are created and with the proper priority. Previously `TestOrderedPolicyValidation()` always set `PolicyEntry.DefaultDeny = true`, added a new option in the test case to disable this to be able to test the case where `PolicyEntry.DefaultDeny == false`.

joestringer commented:
@jrajahalme Is this a user-facing change? Similar to [this comment](https://github.com/cilium/cilium/pull/42784#issuecomment-3726774339) I wonder whether how to appropriately communicate the changes to users. If the user can't use the functionality, I don't think we should elevate the release note in the list. At the same time I think it would be valuable to inform users that there is ongoing work in this area. Maybe we can set to `release-note/misc` but then in the final v1.19.0 announcement we just note that there is ongoing work within the policy engine to prepare for KCNP support?
